merkletree.go

```diff
@@ -0,0 +1,274 @@
+)
+
+const (
+	EmptyBranchIdentifier = 'E'
+	LeafIdentifier        = 'L'
+)
+
+type MerkleTree struct {
+	treeNonce []byte
+	salt      []byte
+	privKey   []byte
+	pubKey    []byte
+	policies  Policies
+	root      *interiorNode
+}
+
+func InitMerkleTree(policies Policies, treeNonce, salt, pubKey, privKey []byte) *MerkleTree {
+	leftBranch := &emptyNode{
+		node: node{
+			level: 1,
+		},
+	}
+
+	rightBranch := &emptyNode{
+		node: node{
+			level: 1,
+		},
+	}
+	root := &interiorNode{
+		node: node{
+			parent: nil,
+			level:  0,
+		},
+		leftChild:  leftBranch,
+		rightChild: rightBranch,
+		leftHash:   nil,
+		rightHash:  nil,
+	}
+	leftBranch.parent = root
+	rightBranch.parent = root
+
+	m := &MerkleTree{
+		treeNonce: treeNonce,
+		salt:      salt,
+		privKey:   privKey,
+		pubKey:    pubKey,
+		policies:  policies,
+		root:      root,
+	}
+	return m
+}
+
+func (h *History) LookUp(key string) (MerkleNode, []LookUpProofNode, error) {
+	str := h.currentSTR
+	return lookUp(key, str)
+}
+
+func (h *History) LookUpInEpoch(key string, ep int64) (MerkleNode, []LookUpProofNode, error) {
+	str := h.GetSTR(ep)
+	return lookUp(key, str)
+}
+
+func lookUp(key string, str *SignedTreeRoot) (MerkleNode, []LookUpProofNode, error) {
+	lookupIndex := computePrivateIndex(key)
+	position := 0
+	var nodePointer interface{}
+	nodePointer = str.treeRoot
+	var proof []LookUpProofNode
+
+	for {
+		if _, ok := nodePointer.(*userLeafNode); ok {
+			// reached to a leaf node
+			break
+		}
+		if _, ok := nodePointer.(*emptyNode); ok {
+			// reached to an empty branch
+			break
+		}
+		proof = append(proof, nodePointer.(*interiorNode))
+		direction := util.GetNthBit(lookupIndex, position)
+		if direction {
+			nodePointer = nodePointer.(*interiorNode).rightChild
+		} else {
+			nodePointer = nodePointer.(*interiorNode).leftChild
+		}
+		position++
+	}
+
+	if nodePointer == nil {
+		return nil, nil, ErrInvalidTree
+	}
+	switch nodePointer.(type) {
+	case *userLeafNode:
+		proof = append(proof, nodePointer.(*userLeafNode))
+		if nodePointer.(*userLeafNode).key == key {
+			return nodePointer.(*userLeafNode), proof, nil
+		}
+		// reached a different leaf with a matching prefix
+		// return nil and a auth path including the leaf node
+		return nil, proof, nil
+	case *emptyNode:
+		proof = append(proof, nodePointer.(*emptyNode))
+		return nil, proof, nil
+	}
+	return nil, nil, ErrInvalidTree
+}
+
+func (m *MerkleTree) Set(key string, value []byte) error {
+	index := computePrivateIndex(key)
+	toAdd := userLeafNode{
+		key:        key,
+		value:      value,
+		index:      index,
+		commitment: crypto.Digest(m.salt, []byte(key), value),
+	}
+
+	return m.insertNode(index, &toAdd)
+}
+
+// Private Index calculation function
+// would be replaced with Ismail's VRF implementation
+func computePrivateIndex(key string) []byte {
+	return crypto.Digest([]byte(key))
+}
+
+func (m *MerkleTree) insertNode(key []byte, toAdd *userLeafNode) error {
```

Comment:
> I should use a consistent naming here. The key parameter is a byte array and the key of toAdd is string. 

Yes, the different types are the result of them not being the same data.

>  I made a separate insertNode since I thought we might have some other methods such as update or things like that in the future. I think it's still fine now.

Fair enough. That's fine.